### DISCUSSION TITLE
Support for SGSR 2; newest temporal upscaler

### DISCUSSION THREAD

**Discussion by Not4ce** [optiscaler/OptiScaler#115] [Ideas]
While we do already have FSR and XeSS, having support for Qualcoms newest TAA based upscaler can provide another flexible upscaler that is faster and closer to image quality to XeSS

**Whilst the question may be: is this worth it? Why rely on mobile tech?**

There are a few good reasons including the desktop focused SGSR2 path which is explained [here](https://github.com/SnapdragonStudios/snapdragon-gsr/tree/main/sgsr/v2). There is also precedence, where the spatial solution SGSR1 already vastly outperforms AMDs mediocre FSR1.1. Due to the recent inclusion of SGSR1 on the lossless scaling app, we can compare it to arguably the best spatial scaler available currently, LS1: https://imgsli.com/MzE4Mzkx

Still images dont tell the whole story. In Motion, SGSR1 is clean with very little blending and smearing, only costing 2ms on a 780m/Rog Ally. This is faster than both FSR1 and Bicubic CAS! It also gets close to offering the image resolve of LS1 which is ML based. This is important, since this tells us SGSR1 is very fast but very competent. This means SGSR2, the TAA based solution may also have a very good image resolve, with less ghosting and fizzle than FSR2

**FSR2; Ghosting, TAAU and SGSR2 desktop pass** 

Qualcomm seemingly wants to do the same for desktop/consoles with a TAA variant which is their 3-pass compute shader. SGSR2 also directly mentions TAAU as competitor with the goal of SGSR2 to have "less ghosting" and better "temporal stability". This is very good, since TAAU can already be better than FSR3/2 (w/ reactive masks) in many games 

The cost as claimed by qualcomm also seems to be lower than FSR2, promising both better image quality and flexibility. Examples show a 2ms render cost for a SD Gen 3 when upscaling by a 1.5x factor from 840x1866 to 1260x2800 which is roughly 3.5m pixel. This may be sub 1ms on any modern desktop graphics card at essentially 1440p, making it faster than FSR2  

If I read their github correctly, it should be possible to take the dlss inputs for SGSR2 and implement it to Optiscaler. This could be a competent alternative to Amds FSR3/2 with a possibility to have less occlusion fizzle and ghosting at a very reasonable compute cost. Even reactive masks don't help on games like Horizon, where FSR has a lot of haloing, grain and smear on textures during camera pan, where a lot of users find those to be much bigger issues than shimmering

![3_pass_implementation](https://github.com/user-attachments/assets/5d0a77c1-3440-4c41-af32-6572d77fee5e)

**Licensing**

Thankfully licensing should be a non issue. As per qualcoms terms here: https://github.com/SnapdragonStudios/snapdragon-gsr/blob/main/LICENSE

They are also fine with modifications, and impose no restrictions on the binaries at stating: "Redistribution and use in source and binary forms, with or without
modification, are permitted"

**Why OptiScaler is important**

I very much appreciate and want to thank anyone and everyone that has worked, & continues to spend time to make this software as good as it is 

With game devopers choosing to not implement basic upscalers, skipping QA, not providing updates and shipping with older versions, Optiscaler has been a game changer

This is very appreciated and important, as games are now designed with upscaling in mind. Games now run poorly, with modern games not targeting native resolutions at all, mandating resolution scaling on modern/UE5 game releases. The cherry on top, dev time is not allocated to confirm if their implemented scalers even work correctly. Either the full featureset is not provided or the FG/scaler has improper frame pacing(fsr3), ghosting or not even work (Cyberpunk, Monster Hunter Wilds, Star Wars, so on). Without Optiscaler, users are locked out of choice or left with broken implementations and subpar image quality 

I may be missing a reason as to why SGSR2 may be tricky to implement or if there are roadmap and logistics behind it; but I would love to know if it was possible or planned. I am unsure what the project pipeline and priorities are, but given what SGSR1 has shown, I would love to have this option as SGSR2 may be a competent & faster alternative to FSR2

This would help a lot of handhelds that are getting popular such as Rog Ally whilst also benefiting Desktops. Having a non-XeSS alternative that is not FSR and faster can help a lot of users reach the 60fps line at 1440p/4k

Personally I think FSR2 looks very poor even scaling from 1440p to 4k, where I simply stick to Xess/TAAU and scale more if needed. Having another choice which allows users to keep a higher res/image quality is what SGSR2 could potentially provide. But this is all based on assumptions and what Qualcomm themselves have outlined as goals when they compare it to TAAU 

Any input regarding this would be appreciated

**Comment by cdozdil** [comment]
Hi,

Thanks for detailed information about SGSR2, it looks interesting. My main issue with it's implementation would be not having a DirectX example or shader (it's using GLSL) and not having a framework to handle resource management, shader setup and dispatch. Because of mobile roots it aims Vulkan/OpenGL ES mostly. I hope it would gather more attention so some implementation examples for DirectX and maybe basic frameworks would pop up. Otherway I am already really busy novadays and might not have find time to implement it soon.

Thanks again for heads up 👍

Edit: Just noticed, looks like their UE plugins have HLSL versions of shaders.

**Reply by Not4ce** [reply]
Hi cdozil, firstly thank you for all the hard work, that is very much appreciated 

And I completely understand if this takes a while and is not a high priority given the scope and time required when working with optiscaler. I was going to mention they have a unreal plugin already, but looks like you already saw that 

I should say if you are sceptical/curious like I was and have lossless scaling, have a quick look at SGSR1 in motion. As someone who had no hope for Qualcomm and their upscaler endeavours, I was very pleasantly surprised with the image resolve. I must admit this was not on my radar at all until I saw it in motion for myself 

Thanks for the response 🙂

**Reply by ninele7** [reply]
Hi, in case someone decides to create a PR with this feature but only with vulkan support would it be acceptable?

**Reply by cdozdil** [reply]
> Hi, in case someone decides to create a PR with this feature but only with vulkan support would it be acceptable?

Sure,

Actually they have HLSL shaders for DirectX too and Maple was checking but after noticing it's a fork of FSR2 and resource management etc will be a bit hassle we decided to wait for SD to release a proper API/Framework.
